submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,327 @@
+
+    // +optional
+    LatencyRTT *LatencyRTTSpec     `json:"latencyRTT,omitempty"`
+}
+
+// ClusterConnectionStatus describe the current state of the ClusterConnectionPolicy.
+type ClusterConnectionStatus struct {
+  // Conditions holds an array of metav1.Condition that describe the state of the ClusterConnectionStatus.
+  Conditions []metav1.Condition `json:"conditions,omitempty"`
+}
+
+// ClusterConnectionPolicyConditionType is the type for status conditions on
+// a ClusterConnectionPolicy This type should be used with the
+// ClusterConnectionStatus.Conditions field.
+type ClusterConnectionConditionType string
+
+const (
+  ClusterConnectionConnected ClusterConnectionConditionType = "Connected"
+  ClusterConnectionNotConnected ClusterConnectionConditionType = "Not Connected"
+  ClusterConnectionError ClusterConnectionConditionType = "Error"
+
+)
+
+//   ClusterConnectionConditionReason defines the set of reasons that explain why a
+// particular ClusterConnectionPolicy condition type has been raised.
+type   ClusterConnectionConditionReason string
+
+const (
+  // ClusterConnectionConditionReasonPolicyConflict represents a reason where the ClusterConnectionPolicy may not have been
+  // implemented in the cluster due to a conflict detected with another ClusterConnectionPolicy
+  ClusterConnectionConditionReasonPolicyConflict  ClusterConnectionConditionReason = "Policy Conflict"
+)
+```
+
+> **_NOTE:_** The `ClusterConnection` CR could be a single CR that aggregates the status of all connections.
+
+### Labelling clusters
+
+Clusters are defined as Custom Resources and can be labeled in the CRD. A new parameter to label clusters at
+`subctl join` time could also be added for usability purposes.
+
+#### Connection Policy Examples
+
+- We want to specify a default policy that uses IPsec as the default cable driver.
+
+```yaml
+kind: ClusterConnectionPolicy
+metadata:
+  name: default
+  namespace: submariner-operator
+spec:
+  cableDriver:
+    name: "ipsec"
+```
+
+- We want to use specific cable drivers to connect clusters in a ClusterSet as shown in the image below. We
+label the `Clusters` appropriately and define a number of policies with the appropriate label selectors.
+
+![Finer grained connectivity example](./images/multiple_cables_ep.png)
+
+```yaml
+kind: ClusterConnectionPolicy
+metadata:
+  name: default
+  namespace: submariner-operator
+spec:
+  cableDriver:
+    name:  "vxlan"
+```
+
+```yaml
+kind: ClusterConnectionPolicy
+metadata:
+  name: on-prem-to-cloud
+  namespace: submariner-operator
+spec:
+  leftClusterSelector:
+    env: cloud
+  rightClusterSelector:
+    env: on-premise
+  cableDriver:
+    name:  "ipsec"
+```
+
+#### Updated Endpoint example
+
+```yaml
+Kind:         Endpoint
+Metadata:
+  Name:         cluster1-submariner-cable-cluster1-172-18-0-14
+  Namespace:    submariner-operator
+Spec:
+  Backends:
+    cable_driver:
+        name: vxlan
+        config:
+          Natt - Discovery - Port:  4490
```

Comment:
I assume you're referring to the `gateway.submariner.io/udp-port` annotation that a user can specify on the gateway node. This is used by the _libreswan_, _vxlan_, and _wireguard_ drivers which would seem to be problematic. I think we'll need to define a udp-port key for each driver.

Reply:
Yeah, these ports are the 'frontend' ports of the GW pod, currently, we have one external udp port (default 4500) that is used for both ipsec and vxlan.

I assume we need to update also the firewall configuration (cloud prepare) in this context (and this should affect also GW K8S LB service when gw pods will run as backends of k8S LBS, which is planned to be supported in the 0.13).

Reply:
Code like [this](https://github.com/submariner-io/submariner/blob/c047698ccbf1fcc43b2dcd7c0773acc902914f61/pkg/routeagent_driver/handlers/kubeproxy/routes_iface.go#L115) in the route agent is also problematic for supporting multiple cable drivers. There may be other hidden assumptions lurking that are tied to having one driver. It seems we need to determine if it's even feasible to run multiple cable drivers in the same cluster and, if so, how.

Reply:
Agree, I think it makes sense first to analyze what components (the ones that assume 1:1 mapping) need to be refactored and refactor these components (I assume that should be a separate Epic)  and after that design&implement new features (like connections policy).

Reply:
I suggest we note these problematic issues in the proposal and address the remaining comments and merge it as in progress so we can pick it up later.

Reply:
> I think we need to have a separate udp port for ipsec and VxLAN , right ?
> 
> Otherwise, we might have a port conflict in case node needs to run both cable drivers.

It looks like ports in the range 4489-4499 could be used https://www.speedguide.net/port.php?port=4490